### ISSUE TITLE
Allow multiple outstanding ICE restarts?

### ISSUE THREAD

**Issue by JonathanLennox** [wish-wg/webrtc-http-ingest-protocol#26]
If you have more than one ice-restart PATCH request in flight at once, it'll be unclear to the server which one was sent first.

This ambiguity is bad; it won't be clear which selected ICE candidate pair to switch to.

One simple solution would simply be to forbid this, but Sergio pointed out at the meeting that this might be problematic if your device fell off its network (because it might not get the response to the first ice-restart, so would have to wait for it to time out).

If so we need another way to resolve the ambiguity.

**Comment by murillo128** [comment]
Would sending a `Last-Modified` header (rfc7232) on the response help solving the issue?

**Comment by murillo128** [comment]
Right, the `Last-Modified` timestamp has only second accuracy, so it would not work if both patch requests have been received in the same second.

This is however kind of an edge case in an edge case. Could we just state that if both ice restarts are received with same `Last-Modified` the client should send a new ice restart request? All the other alternatives seems uglier and more complex.

**Comment by JonathanLennox** [comment]
Basing it on "if two were received" doesn't let the server know which one the client is using, in the case where one of the responses wasn't received (because the whole mess was triggered by a network failure).

However, in that case one of the ICE handshakes won't succeed, so the server can just continue with the (hopefully) succeeding one.

I agree this is an edge case of an edge case.

**Comment by murillo128** [comment]
I think I have worded my comment badly. Let's back and see which would be the problematic case.

The client will update its ICE username fragment and pwd before sending the PATH request and the server will restart ice when receiving a new PATCH request with an updated ice username fragment/pwd.

I don't think that there is any issue at all if any of the requests or responses are not received by either side.

The issue happens when client has sent two PATCH requests (first with username client1 and second with client2) and the server receives them out of order and generates two responses one with server 1 and the later with server 2. So the client would be using client2:server1 and the server server2:client1. 

The `Last-Modified` header would allow us to know (up to a 1s of resolution) which request has been processed first by the server, but would not allow us to solve the problem, and the client would have to send a new Ice restart request in order to get to a good state. 

The other alternative is always send an `If-Match` with the last know etag, that would prevent the server for processing out of order ice restart request, but it would prevent the ice restart from working if the previous response was lost and we don't know the value of the actual server etag.

**Comment by jech** [comment]
Let us please not use `Last-Modified` in new protocols.  It causes a number of complications which implementers are likely to get wrong (dealing with sub-second transactions, interaction with `Vary`, etc.).  Let's please stick to ETags only.

**Comment by jech** [comment]
Just thinking aloud...

Send PATCH with If-Match.  If you get 412, resynchronise using HEAD and try again?

**Comment by JonathanLennox** [comment]
> Just thinking aloud...
> 
> Send PATCH with If-Match. If you get 412, resynchronise using HEAD and try again?

If-Match with which ETag?  Remember the worrying case is where you send the second request before the response to the first request has arrived.  (And yes, if interfaces are flapping, it's possible for it to turn out that there are to be *three* (or more) outstanding requests...)

**Comment by jech** [comment]
> Remember the worrying case is where you send the second request before the response to the first request has arrived.

Ah, I see.  Client changes attachment, reply never arrives, client triggers ICE restart.

**Comment by JonathanLennox** [comment]
> > Remember the worrying case is where you send the second request before the response to the first request has arrived.
> 
> The last one the client got. If it's obsolete (which is likely after a failrure), client gets 412, gets the last ETag using HEAD, then resends with the right ETag.

That means (at least) 3 RTTs to recover from a network failure, which isn't great, but I agree it's probably the easiest solution protocol-wise.  A double ICE restart like this should hopefully be a pretty rare phenomenon.

**Comment by jech** [comment]
@JonathanLennox Agreed on both counts.

**Comment by murillo128** [comment]
How about not making the protocol harder and let the client deal with it by sending a new ICE restart in case there are receive an out of order response and ICE is not in `connected` state?

**Comment by murillo128** [comment]
@JonathanLennox @jech what do you think about my previous comment?

**Comment by JonathanLennox** [comment]
@murillo128 The problem is that receiving a response out of order doesn't necessarily mean that that request was processed out of order, and vice-versa.  Also, the out-of-order response might not be received, because this will be happening at times of network flakiness.

I like @jech's suggestion, but it would mean defining HEAD (and thus presumably GET) on the WHIP resource URL in order to get resync working, which isn't great.

**Comment by murillo128** [comment]
I have re-read my previous comment, and it does not make any sense in fact. Let me try again to rephrase my thoughts.

The server response order is irrelevant. When the client performs an ICE restart, it discards the ice username/pwd frags, so if there was a previous request pending to receive an answer it must be discarded. So only the response to the last sent request is relevant to the client. So the key question is detecting out of order request at the server. 

Instead of trying to resolve it at HTTP level, we could just apply the ICE info received by the server and wait for the ICE connection states. If there is a mismatch between the information at the client and at the server (because an out of order request), the STUN request will contain an invalid ICE info and will be rejected by the server. 

In terms of webrtc w3c apis, this would trigger an ice `disconnected`  or `failed` which would be an indication for the client to perform a new ICE restart. 

Depending on the local candidates, it may require the ice request to timeout (30s?) but it would be the client would eventually recover. The client could also optimize this by adding a timeout to receive the `completed` state to start a new ice restart (as already done by many non trickle ice services to not have to wait for candidates in private networks).

**Comment by jech** [comment]
So you're arguing that we make the ICE restart best-effort, and rely on the application to trigger a new restart if the previous restart failed.  That makes a lot of sense.  However, I'm a little concerned this will cause unacceptable latency (see RFC 8863).

**Comment by murillo128** [comment]
We are talking about an edge case of an edge case on unstable networks.. best effort is more than enough IMHO.. the alternative is complicating *a lot* the protocol..

**Comment by jech** [comment]
I reluctantly agree.  I still think the failure should be signalled explicitly at the HTTP layer, at the very least it will make troubleshooting easier.

**Comment by murillo128** [comment]
@JonathanLennox @jech wdyt?

**Comment by jech** [comment]
Why not require that the server indicate the mismatch to the client, e.g. with HTTP 409?

**Comment by murillo128** [comment]
because the server has no way of detecting the out of order requests

**Comment by jech** [comment]
I'll switch my brain on next time.  Sorry.

**Comment by murillo128** [comment]
@jech @JonathanLennox any feedback on this? I would love to update the draft asap and to a WGLC on it..

**Comment by jech** [comment]
I feel we should go into LC with the issue open.

**Comment by JonathanLennox** [comment]
It sounds like I'm in the rough on making the protocol more complex, so if you want to write up your solution as a PR I'm willing to review it to make sure it makes sense.

**Comment by murillo128** [comment]
@JonathanLennox the PR is in here: https://github.com/wish-wg/webrtc-http-ingest-protocol/pull/32
